pkg/operator/finalizer.go

```diff
@@ -0,0 +1,167 @@
+package operator
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	regopapi "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1alpha1"
+	"github.com/openshift/cluster-image-registry-operator/pkg/generate"
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+)
+
+func (h *Handler) RemoveResources(o *regopapi.ImageRegistry) error {
+	modified := false
+
+	templetes, err := h.GenerateTemplates(o, &h.params)
+	if err != nil {
+		return fmt.Errorf("unable to genetate templates: %s", err)
+	}
+
+	for _, tmpl := range templetes {
+		err = generate.RemoveByTemplate(tmpl, &modified)
+		if err != nil {
+			return fmt.Errorf("unable to remove objects: %s", err)
+		}
+		logrus.Infof("resource %s removed", tmpl.Name())
+	}
+
+	configState, err := generate.GetConfigState(h.params.Deployment.Namespace)
+	if err != nil {
+		return fmt.Errorf("unable to get previous config state: %s", err)
+	}
+
+	err = generate.RemoveConfigState(configState)
+	if err != nil {
+		return fmt.Errorf("unable to remove previous config state: %s", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) finalizeResources(o *regopapi.ImageRegistry) error {
+	defer utilruntime.HandleCrash()
+
+	if o.ObjectMeta.DeletionTimestamp == nil {
+		return nil
+	}
+
+	finalizers := []string{}
+	for _, v := range o.ObjectMeta.Finalizers {
+		if v != parameters.ImageRegistryOperatorResourceFinalizer {
+			finalizers = append(finalizers, v)
+		}
+	}
+
+	if len(finalizers) == len(o.ObjectMeta.Finalizers) {
+		return nil
+	}
+
+	logrus.Infof("finalizing %s %s/%s", o.GetObjectKind().GroupVersionKind().Kind, o.Namespace, o.Name)
+
+	err := h.RemoveResources(o)
+	if err != nil {
+		return err
+	}
+
+	cr := o
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if cr == nil {
+			cr = &regopapi.ImageRegistry{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: o.TypeMeta.APIVersion,
+					Kind:       o.TypeMeta.Kind,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      o.ObjectMeta.Name,
+					Namespace: o.ObjectMeta.Namespace,
+				},
+			}
+
+			err := sdk.Get(cr)
+			if err != nil {
+				return fmt.Errorf("failed to get custom resource %s/%s: %s", o.ObjectMeta.Namespace, o.ObjectMeta.Name, err)
+			}
+		}
+
+		cr.ObjectMeta.Finalizers = finalizers
```

Comment:
I don't think you can do this.  What if the update that caused the conflict was someone else adding a finalizer?  Now you'll have removed their finalizer.

Reply:
Hm. How I should remove my finalizer ?

Reply:
i think you need to iterate the list again and remove it again, since the list may have changed.

Reply:
(that is, the same thing you're doing on line 59)

Reply:
in practice? probably no one will.  But in theory, anyone w/ sufficient permission could and i'd rather not leave a potential hole open in case we do later introduce another finalizer for some reason and had forgotten about this.

Reply:
Why then do we not follow the situation specifying our operator as an initializer (in theory this is also possible)?

Reply:
What would that accomplish for us?

Reply:
@legionus they are mainly made for mutating/validating admission controllers.